### PR TITLE
Cap number of reduce tasks used by Hive

### PR DESCRIPTION
**Pull request by mulby**
@brianhw

### CODE REVIEW

**Comment by brianhw** [comment]
I've not looked carefully, but are there other jobs further "downstream" that need to be plumbed to pass this through?

File: edx/analytics/tasks/util/hive.py [hunk 1]
```diff
@@ -45,6 +45,18 @@ class HiveTableTask(WarehouseMixin, OverwriteOutputMixin, HiveQueryTask):
     until they are cleaned up by some external process.
     """
 
+    n_reduce_tasks = luigi.Parameter(default=25, significant=False)
+
+    def hiveconfs(self):
+        jcs = super(HiveTableTask, self).hiveconfs()
+
+        # Don't force the number of reducer to a particular number, instead set a max number of reducers.
```
Comment (brianhw) [inline edx/analytics/tasks/util/hive.py:53]: s/reducer/reducers/

File: edx/analytics/tasks/util/hive.py [hunk 2]
```diff
@@ -45,6 +45,18 @@ class HiveTableTask(WarehouseMixin, OverwriteOutputMixin, HiveQueryTask):
     until they are cleaned up by some external process.
     """
 
+    n_reduce_tasks = luigi.Parameter(default=25, significant=False)
+
+    def hiveconfs(self):
+        jcs = super(HiveTableTask, self).hiveconfs()
```
Comment (brianhw) [inline edx/analytics/tasks/util/hive.py:51]: Please use a more readable variable name than jcs.

File: edx/analytics/tasks/util/hive.py [hunk 3]
```diff
@@ -45,6 +45,18 @@ class HiveTableTask(WarehouseMixin, OverwriteOutputMixin, HiveQueryTask):
     until they are cleaned up by some external process.
     """
 
+    n_reduce_tasks = luigi.Parameter(default=25, significant=False)
```
Comment (brianhw) [inline edx/analytics/tasks/util/hive.py:48]: If this is really max_reduce_tasks, shouldn't it be so named?

File: edx/analytics/tasks/util/hive.py [hunk 4]
```diff
@@ -45,6 +45,18 @@ class HiveTableTask(WarehouseMixin, OverwriteOutputMixin, HiveQueryTask):
...
+
+    def hiveconfs(self):
+        jcs = super(HiveTableTask, self).hiveconfs()
+
+        # Don't force the number of reducer to a particular number, instead set a max number of reducers.
+        # For jobs with small numbers of input groups, this will perform better since it won't have to spin up a lot of
+        # reducers to do a small amount of work.
+        num_reduce_tasks = jcs.pop('mapred.reduce.tasks', self.n_reduce_tasks)
```
Comment (brianhw) [inline edx/analytics/tasks/util/hive.py:56]: max_reduce_tasks?

**Comment by brianhw** [comment]
Okay, best I can tell, all the downstream classes of HiveQueryToMysqlTask are passing in n_reduce_tasks, including classes derived from CourseActivityTask and EnrollmentTask.

**Comment by mulby** [comment]
Yeah I have been just hacking on this to see what the performance impact is. I agree that a) jcs needs to be renamed and b) overloading n_reduce_tasks like this is confusing and we should do something more clear OR just use it the way luigi intended and have it set a fixed number of reducers instead of the max.

**Comment by mulby** [comment]
@brianhw

**Comment by brianhw** [comment]
Wasn't sure what the next step on this is.  I had the impression that the performance gain was small, so that you were inclined to not pursue this.  But let me know if that assumption was wrong.

**Comment by mulby** [comment]
That's correct, closing.

**Pull request closed without merging**